snuba/query/processors/mapping_optimizer.py

```diff
@@ -159,20 +185,110 @@ def __replace_with_hash(self, condition: Expression) -> Expression:
             ),
         )
 
+    def _get_condition_without_redundant_checks(
+        self, condition: Expression, query: Query
+    ) -> Expression:
+        """Optimizes the case where the query condition contains the following:
+
+        valueOf('my_tag') != '' AND valueOf('my_tag') == "something"
+                          ^                            ^
+                          |                            |
+                      existence check               value check
+
+        the existence check in this clause is redundant and prevents the hashmap
+        optimization from being applied.
+
+        This function will remove all tag existence checks
+        from the condition IFF they are ANDed with a value check for the *same tag name*
+
+        Side effects:
+            This function works by flattening first level AND conditions to find clauses where
+            existence checks and value checks are ANDed together. When the AND conditions are recombined,
+            they are not guaranteed to be in the same structure (but are guaranteed to be functionally equivalent)
+
+            Example:
+                ┌───┐         ┌───┐
+                │AND│         │AND│
+                ├──┬┘         └┬──┤
+                │  │           │  │
+             ┌──┴┐ c           a ┌┴──┐
+             │AND│    becomes    │AND│
+             └┬─┬┘               ├──┬┘
+              │ │                │  │
+              a b                b  c
+        """
+        if not isinstance(condition, FunctionExpr):
+            return condition
+        elif condition.function_name == BooleanFunctions.OR:
+            sub_conditions = get_first_level_or_conditions(condition)
+            pruned_conditions = [
+                self._get_condition_without_redundant_checks(c, query)
+                for c in sub_conditions
+            ]
+            return combine_or_conditions(pruned_conditions)
+        elif condition.function_name == BooleanFunctions.AND:
+            sub_conditions = get_first_level_and_conditions(condition)
+            tag_eq_match_strings = set()
+            matched_tag_exists_conditions = {}
+            for condition_id, cond in enumerate(sub_conditions):
+                tag_exist_match = _tag_exists_pattern.match(cond)
+                if tag_exist_match:
+                    matched_tag_exists_conditions[condition_id] = tag_exist_match
+                eq_match = self.__optimizable_pattern.match(cond)
```

Comment:
put this into an else? if `tag_exists_match` is true this cannot be true so you can skip the check. This may have an impact as it happens really often.